libs/MeshKernel/src/CasulliDeRefinement.cpp

```diff
@@ -0,0 +1,1020 @@
+// the Free Software Foundation version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+// contact: delft3d.support@deltares.nl
+// Stichting Deltares
+// P.O. Box 177
+// 2600 MH Delft, The Netherlands
+//
+// All indications and logos of, and references to, "Delft3D" and "Deltares"
+// are registered trademarks of Stichting Deltares, and remain the property of
+// Stichting Deltares. All rights reserved.
+//
+//------------------------------------------------------------------------------
+
+#include <algorithm>
+
+#include "MeshKernel/CasulliDeRefinement.hpp"
+#include "MeshKernel/Exceptions.hpp"
+#include "MeshKernel/Operations.hpp"
+#include "MeshKernel/UndoActions/AddNodeAction.hpp"
+#include "MeshKernel/UndoActions/CompoundUndoAction.hpp"
+#include "MeshKernel/UndoActions/FullUnstructuredGridUndo.hpp"
+
+std::unique_ptr<meshkernel::UndoAction> meshkernel::CasulliDeRefinement::Compute(Mesh2D& mesh) const
+{
+    Polygons emptyPolygon;
+    return Compute(mesh, emptyPolygon);
+}
+
+std::unique_ptr<meshkernel::UndoAction> meshkernel::CasulliDeRefinement::Compute(Mesh2D& mesh, const Polygons& polygon) const
+{
+    std::unique_ptr<FullUnstructuredGridUndo> refinementAction = FullUnstructuredGridUndo::Create(mesh);
+    const std::vector<Point> meshNodes(mesh.Nodes());
+    const std::vector<Edge> meshEdges(mesh.Edges());
+
+    if (DoDeRefinement(mesh, polygon))
+    {
+        mesh.DeleteInvalidNodesAndEdges();
+        mesh.Administrate();
+    }
+    else
+    {
+        // The de-refinement failed, restore the original mesh
+        refinementAction->Restore();
+        // Reset the undo action to a null action
+        refinementAction.reset(nullptr);
+        throw AlgorithmError("Unable to compute the Casulli de-refinement");
+    }
+
+    return refinementAction;
+}
+
+void meshkernel::CasulliDeRefinement::FindDirectlyConnectedCells(const Mesh2D& mesh,
+                                                                 const UInt elementId,
+                                                                 std::vector<UInt>& connected) const
+{
+    connected.clear();
+
+    for (UInt i = 0; i < mesh.m_numFacesNodes[elementId]; ++i)
+    {
+        UInt edgeId = mesh.m_facesEdges[elementId][i];
+
+        if (mesh.m_edgesNumFaces[edgeId] < 2)
+        {
+            continue;
+        }
+
+        UInt neighbouringElementId = elementId == mesh.m_edgesFaces[edgeId][0] ? mesh.m_edgesFaces[edgeId][1] : mesh.m_edgesFaces[edgeId][0];
+
+        if (std::ranges::find(connected, neighbouringElementId) == connected.end())
```

Comment:
do you need this check, std::ranges::find for every added face? it looks like you are simply looping over the edges of the faces. the added faces should not be the same if there are no hanging nodes